cmd/juju/commands/bundle.go

```diff
@@ -111,7 +112,50 @@ func (h *bundleHandler) addCharm(id string, p bundlechanges.AddCharmParams) erro
 // addService deploys or update a service with no units. Service options are
 // also set or updated.
 func (h *bundleHandler) addService(id string, p bundlechanges.AddServiceParams) error {
-	// TODO frankban: implement this method.
+	status, err := h.client.Status(nil)
+	if err != nil {
+		return errors.Annotate(err, "cannot retrieve environment status")
+	}
+	svcStatus, svcExists := status.Services[p.Service]
+	// TODO frankban: the charm should really be resolved using
+	// resolve(p.Charm, h.results) at this point: see TODO in addCharm.
+	ch := h.results["resolved-"+p.Charm]
+	if svcExists {
+		// The service is already deployed in the environment: check that its
+		// charm is compatible with the one declared in the bundle. If it is,
+		// reuse the existing service or upgrade to a specified revision.
+		// Exit with an error otherwise.
+		if svcStatus.Charm == ch {
+			h.log.Infof("reusing service %s (charm: %s)", p.Service, ch)
+		} else {
+			if err := checkCompatibleCharms(ch, svcStatus.Charm); err != nil {
+				return errors.Annotatef(err, "cannot upgrade charm for service %q", p.Service)
+			}
+			h.log.Infof("upgrading charm for existing service %s (from %s to %s)", p.Service, svcStatus.Charm, ch)
+			if err := h.client.ServiceSetCharm(p.Service, ch, false); err != nil {
+				return errors.Annotatef(err, "cannot upgrade charm for service %q", p.Service)
+			}
+		}
+	} else {
+		// The service does not exist in the environment.
+		h.log.Infof("deploying service %s (charm: %s)", p.Service, ch)
+		// TODO frankban: handle service constraints in the bundle changes.
+		numUnits, configYAML, cons, toMachineSpec := 0, "", constraints.Value{}, ""
+		if err := h.client.ServiceDeploy(ch, p.Service, numUnits, configYAML, cons, toMachineSpec); err != nil {
+			return errors.Annotatef(err, "cannot deploy service %q", p.Service)
+		}
+	}
+	if len(p.Options) > 0 {
+		h.log.Infof("configuring service %s", p.Service)
+		config, err := yaml.Marshal(map[string]map[string]interface{}{p.Service: p.Options})
```

Comment:
Should this go out in a helper method so that we separate input data handling from the addService logic?